cleanlab/experimental/datalab/issue_manager/duplicate.py

```diff
@@ -0,0 +1,147 @@
+            self.knn = NearestNeighbors(n_neighbors=self.k, metric=self.metric)
+
+        try:
+            check_is_fitted(self.knn)
+        except:
+            self.knn.fit(feature_array)
+
+        scores, distances = self._score_features(feature_array)
+        self.radius, self.threshold = self._compute_threshold_and_radius()
+
+        self.issues = pd.DataFrame(
+            {
+                f"is_{self.issue_name}_issue": scores < self.threshold,
+                self.issue_score_key: scores,
+            },
+        )
+
+        indices = self.knn.radius_neighbors(feature_array, self.radius, return_distance=False)
+        self.near_duplicate_sets = [
+            duplicates[duplicates != idx] for idx, duplicates in enumerate(indices)
+        ]
+        self.distances = distances
+
+        self.summary = self.get_summary(score=scores.mean())
+        self.info = self.collect_info()
+
+    def collect_info(self) -> dict:
+        issues_dict = {
+            "num_near_duplicate_issues": len(self.near_duplicate_sets),
+            "average_near_duplicate_score": self.issues[self.issue_score_key].mean(),
+            "near_duplicate_sets": self.near_duplicate_sets,
+            "radius": self.radius,
+        }
+
+        params_dict = {
+            "metric": self.metric,
+            "k": self.k,
+            "threshold": self.threshold,
+        }
+
+        knn = self.knn
+        dists, nn_ids = [array[:, 0] for array in knn.kneighbors()]  # type: ignore[union-attr]
+        weighted_knn_graph = knn.kneighbors_graph(mode="distance").toarray()  # type: ignore[union-attr]
+
+        # TODO: Reverse the order of the calls to knn.kneighbors() and knn.kneighbors_graph()
+        #   to avoid computing the (distance, id) pairs twice.
+        knn_info_dict = {
+            "nearest_neighbour": nn_ids.tolist(),
+            "distance_to_nearest_neighbour": dists.tolist(),
+            # TODO Check scipy-dependency
+            "weighted_knn_graph": weighted_knn_graph.tolist(),
+        }
+
+        info_dict = {
+            **issues_dict,
+            **params_dict,
+            **knn_info_dict,
+        }
+        return info_dict
+
+    def _extract_embeddings(self, columns: Union[str, List[str]], **kwargs) -> np.ndarray:
+        """Extracts embeddings for the given columns."""
+
+        if isinstance(columns, list):
+            raise NotImplementedError("TODO: Support list of columns.")
+
+        format_kwargs = kwargs.get("format_kwargs", {})
+
+        return self.datalab.data.with_format("numpy", **format_kwargs)[columns]
+
+    def _score_features(self, feature_array) -> Tuple[np.ndarray, np.ndarray]:
+        """Computes nearest-neighbor distances and near-duplicate scores for input features"""
+        knn = cast(NearestNeighbors, self.knn)
+        distances, _ = knn.kneighbors(feature_array)
+        distances = distances[:, 1]  # nearest neighbor is always itself
+
+        self.distances = distances
```

Comment:
`self.distances` is assigned here and then again in [line 65](https://github.com/cleanlab/cleanlab/pull/614/files#diff-62e85fda10b03f3cd91084bd749f3822834ec6242184eabc62668950f9b4b422R65)

Reply:
Good eye! I've decided to keep the attribute assignment in `find_issues`.
I do the assignment earlier in that method now so things don't break.